src/NUnitEngine/nunit-agent/Program.cs

```diff
@@ -130,58 +122,65 @@ public static void Main(string[] args)
             log.Info("Initializing Services");
             engine.Initialize();
 
-            log.Info("Starting RemoteTestAgent");
-            Agent = new RemoteTestAgent(AgentId, AgencyUrl, engine.Services);
-
+            log.Info("Starting AgentServer");
             try
             {
-                if (Agent.Start())
-                    WaitForStop();
-                else
+                using (var server = AgentServer.Start(IPAddress.Loopback, engine.Services.GetService<ITestRunnerFactory>()))
                 {
-                    log.Error("Failed to start RemoteTestAgent");
-                    Environment.Exit(AgentExitCodes.FAILED_TO_START_REMOTE_AGENT);
+                    // Write to the console to let the parent process know which port Windows assigned
+                    Console.WriteLine("Listening on " + server.ListeningOn);
+                    log.Info("Listening on " + server.ListeningOn);
+
+                    Console.WriteLine("Type 'stop' to exit.");
+
+                    while (true)
+                    {
+                        var line = Console.ReadLine();
+                        if (line is null)
+                        {
+                            if (parentProcess.HasExited)
+                            {
+                                log.Error("Parent process has been terminated.");
+                                return AgentExitCodes.PARENT_PROCESS_TERMINATED;
+                            }
+
+                            log.Info("Stopping AgentServer in response to standard input being closed.");
+                            break;
+                        }
+
+                        if ("stop".Equals(line, StringComparison.OrdinalIgnoreCase))
+                        {
+                            log.Info("Stopping AgentServer in response to 'stop' command from standard input.");
+                            break;
+                        }
+
+                        log.Info($"Ignoring unrecognized command '{line}' from standard input.");
+                        Console.WriteLine($"Unrecognized command '{line}'.");
+                    }
                 }
             }
             catch (Exception ex)
             {
-                log.Error("Exception in RemoteTestAgent", ex);
-                Environment.Exit(AgentExitCodes.UNEXPECTED_EXCEPTION);
+                log.Error("Unexpected exception", ex);
```

Comment:
Appreciate this is existing code - I think we might want to use `ExceptionHelper.Flatten()` or something here - make sure we do get all the logging info if things go wrong!

Reply:
I'm not sure where `ExceptionHelper.Flatten` lives but I'm happy to copy that over. Is that the framework repo?

Reply:
Sorry - it's ExceptionHelper.BuildMessage/BuildStackTrace etc - I'd mis-remembered the names. (Flatten is the private methods in that class instead!)

It should actually probably be baked lower into the Logger classes I guess...probably out of scope for this PR. Let's leave this as it. 🙂